daemons/based/based_transaction.c

```diff
@@ -0,0 +1,380 @@
+
+        rc = pcmk_legacy2rc(rc);
+        if (rc != pcmk_rc_ok) {
+            crm_err("Aborting CIB transaction for client %s (%s) due to failed "
+                    "%s request: %s",
+                    client_name, client_id, op, pcmk_rc_str(rc));
+            crm_log_xml_info(data->request, "Failed request");
+            free_request_data(data);
+            return rc;
+        }
+
+        crm_trace("Applied %s request to transaction working CIB for client %s "
+                  "(%s)",
+                  op, client_name, client_id);
+        crm_log_xml_trace(data->request, "Successful request");
+        free_request_data(data);
+    }
+
+    return pcmk_rc_ok;
+}
+
+/*!
+ * \internal
+ * \brief Commit a given CIB client's transaction to a working CIB copy
+ *
+ * \param[in]     client      Client whose transaction to commit
+ * \param[in,out] result_cib  Where to store result CIB
+ *
+ * \return Standard Pacemaker return code
+ *
+ * \note This function is expected to be called only by
+ *       \p cib_process_commit_transaction().
+ * \note \p result_cib is expected to be a copy of the current CIB with change
+ *       tracking enabled, as created by \p cib_perform_op().
+ * \note The caller is responsible for activating and syncing \p result_cib on
+ *       success, and for freeing it on failure.
+ */
+int
+based_commit_transaction(const pcmk__client_t *client, xmlNodePtr *result_cib)
+{
+    GQueue *transaction = NULL;
+    const char *client_name = NULL;
+    const char *client_id = NULL;
+    xmlNodePtr saved_cib = the_cib;
+    int rc = pcmk_rc_ok;
+
+    CRM_ASSERT((client != NULL) && (result_cib != NULL));
+
+    /* *result_cib should be a copy of the_cib (created by cib_perform_op()). If
+     * not, make a copy now. Change tracking isn't required because:
+     * * Each request in the transaction will have changes tracked and ACLs
+     *   checked if appropriate.
+     * * cib_perform_op() will infer changes for the commit request at the end.
+     */
+    CRM_CHECK((*result_cib != NULL) && (*result_cib != the_cib),
+              *result_cib = copy_xml(the_cib));
+
+    transaction = get_transaction(client);
+    if (transaction == NULL) {
+        return pcmk_rc_no_transaction;
+    }
+
+    client_name = pcmk__s(client->name, "unspecified");
+    client_id = pcmk__s(client->id, "unknown");
+
+    crm_trace("Committing transaction for client %s (%s)",
+              client_id, client_name);
+
+    // Apply all changes to a working copy of the CIB
+    the_cib = *result_cib;
+
+    rc = process_transaction_requests(transaction, client_name, client_id);
+
+    if (rc == pcmk_rc_ok) {
+        crm_trace("Transaction successfully committed to working CIB for "
+                  "client %s (%s)",
+                  client_name, client_id);
+    }
+
+    crm_trace("Finished processing transaction requests for client %s (%s); "
+              "discarding request queue",
+              client_name, client_id);
```

Comment:
I prefer to combine messages when possible since logs can get mind-numbingly large e.g. `(rc == pcmk_rc_ok)? "Successfully" : "Unsuccessfully"`

Reply:
Good point. I added the "Finished processing" message later to explain why the "Discarding" message was happening.

Maybe "Transaction commit [succeeded|failed] for client `<name>` (`<id>`); discarding queue". We already have the "aborted" message with the RC string at error level if it failed.

Reply:
sounds good